File: go/procedural/hash.go
package main

import (
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"strings"
)

func generateSalt(length int) (string, error) {
	bytes := make([]byte, length)
	if _, err := rand.Read(bytes); err != nil {
		return "", err
	}
	return hex.EncodeToString(bytes), nil
}

func simpleXORHash(input string, key byte) string {
	var hash strings.Builder
	for _, b := range []byte(input) {
		hashedByte := b ^ key
		hash.WriteString(fmt.Sprintf("%02x", hashedByte))
	}
	return hash.String()
}

func main() {
	password := "password123"
	salt, err := generateSalt(16)
	if err != nil {
		panic(err)
	}

	saltedInput := password + salt
	hash := simpleXORHash(saltedInput, 0xAA)

	fmt.Printf("Salt: %s\nHash: %s\n", salt, hash)
}

File: rust/procedural/hash.rs
use rand::Rng;
use std::fmt::Write;

fn generate_salt(length: usize) -> String {
    let mut rng = rand::thread_rng();
    let mut salt = String::with_capacity(length * 2);
    
    for _ in 0..length {
        let byte: u8 = rng.r#gen();
        write!(&mut salt, "{:02x}", byte).unwrap();
    }
    salt
}

fn simple_xor_hash(input: &str, key: u8) -> String {
    let mut hash = String::with_capacity(input.len() * 2);
    
    for c in input.chars() {
        let hashed_byte = c as u8 ^ key;
        write!(&mut hash, "{:02x}", hashed_byte).unwrap();
    }
    hash
}

fn main() {
    let password = "password123";
    let salt = generate_salt(16);
    let salted_input = format!("{}{}", password, salt);
    let hash = simple_xor_hash(&salted_input, 0xAA);
    
    println!("Salt: {}\nHash: {}", salt, hash);
}